Improve auction administration

adelevie opened the pull request:
These fields will make it easier to administer auctions.

This branch include @baccigalupi's `skinner user junks` and auction types code (thanks, Kane!).

Still WIP.

adelevie commented:
I rebased off the work that @baccigalupi started. That's why it says this entire commit was from Kane. Any bad code in there was from me, though, and not Kane.

jessieay commented:
@adelevie there are merge conflicts

adelevie commented:
@jessieay I just rebased and force pushed. not sure why the button hasn't turned green :/

jessieay commented:
@adelevie are you sure you're rebasing against origin master and not local master?

adelevie commented:
@jessieay that was it. I rebased against stale develop. I just pulled latest and am in the middle of rebasing again.

adelevie commented:
rebased, @jessieay

jessieay commented:
@adelevie what is the "why" behind this Pr? Happy to review code but easier to think about alternative implementations / edge cases if I understand the reasons behind this PR. Usually in a PR description I try to answer these two questions:

Why is this change necessary?
How does it address the issue?

adelevie commented:
great questions, @jessieay.

> Why is this change necessary?

The app was doing a good enough job of letting vendors bid on auctions, but all post-auction administration occurred outside of the app. The administration included a lot of small bits of work that were easy to lose track of.

For example, we never tracked in the app the delivery deadlines, so there was no canonical way to check if any vendors had let the deadline lapse (or if they were about to). The old way was to refer to confirmation emails that I manually emailed to the vendors. I will still do that, but thanks to this PR, we store the delivery deadline in the database.

In this PR, we can check other auction states that were previously stored in my brain or inside of gmail or pencil and paper:

- whether an auction is complete and successful
- whether the winner of an auction needs to be paid
- whether an admin needs to review a pull request from a vendor
- once a vendor's work has been accepted, who do I bill? (we know now thanks to the `billable_to` column on `auctions`).
- has the vendor's payment gone through? (with this PR, we store `cap_proposal_url`. Once I fill out the CAP form, I save the link onto the auction in the admin panel. Whenever a vendor requests, or I feel like it, I can quickly check the payment status. I've run into this a couple times and having this feature will help ensure that no paymeny falls through the cracks).

Once the requisite columns were added, I built into the admin panel some filtered queries for auctions in different states. In `/admin/action_items`, I can view lists of auctions at different stages in their lifecycle and click "edit" to move them through the pipeline (e.g. by clicking a form select box and clicking 'paid').

There are two more smaller reasons for this PR:

- added a feature to draft an auction and preview it without publishing it. Why? Because we've been storing drafts in Google Docs and I very quickly lost track of the different states of the drafts (who needs to do what with this doc?) 
- added a `credit_card_form_url` to the `users` table. This info is not yet enterable via the admin panel (but will be in a subsequent PR). 

Many of the extra fields added in this PR will be used to make POST requests to the CAP API. There is no write API yet, but I am going to work on one, and I want to store the data necessary to programmatically submit payment requests.

> How does it address the issue?

Mostly through migrations and the creation of an `AuctionQuery` class. `AuctionQuery` lets us use composeable scopes without junking up the `Auction` model. If you look at the file, you'll notice a series of 3 line methods defining scopes and some longer methods which are compositions of the smaller scopes.

Other ways the issue is addressed include new templates and updates to controllers.

allalala commented:
A great pr description, @adelevie! :clap: :clap:

harrisj commented on spec/controllers/bids_controller_spec.rb:
```diff
@@ -43,6 +62,24 @@
+
+        it 'renders the template' do
+          get :new, auction_id: auction.id
+          expect(response).to redirect_to("/login")
```
These tests are not doing what the `it` text says they should. Minor nit though

harrisj commented:
These are a lot of changes, but I've muddled through it and it seems like it's all good. Looks good to me for merging
